src/docfx/build/xref/ExternalXrefMapLoader.cs

```diff
@@ -37,47 +41,71 @@ internal class ExternalXrefMapLoader
+                            spec.DocsetName = xrefMap.DocsetName;
+                            externalXrefMap.TryAdd(spec.Uid, new Lazy<ExternalXrefSpec>(() => spec));
                         }
+                        externalXref.AddRange(xrefMap.ExternalXrefs);
                     }
                     else
                     {
+                        var externalXrefSpecAndExternalXref = LoadJsonFile(physicalPath);
+
                         // Fast pass for JSON xref files
-                        foreach (var (uid, spec) in LoadJsonFile(physicalPath))
+                        foreach (var (uid, spec) in externalXrefSpecAndExternalXref.externalXrefSpec)
                         {
                             // for same uid with multiple specs, we should respect the order of the list
-                            result.TryAdd(uid, spec);
+                            externalXrefMap.TryAdd(uid, spec);
                         }
+                        externalXref = externalXrefSpecAndExternalXref.externalXref;
                     }
                 }
 
-                result.TrimExcess();
-                return result;
+                externalXrefMap.TrimExcess();
+                externalXref.TrimExcess();
+
+                return new ExternalXrefMap(externalXrefMap, externalXref);
             }
         }
 
-        public static List<(string, Lazy<ExternalXrefSpec>)> LoadJsonFile(string filePath)
+        public static (List<(string, Lazy<ExternalXrefSpec>)> externalXrefSpec, List<ExternalXref> externalXref) LoadJsonFile(string filePath)
         {
-            var result = new List<(string, Lazy<ExternalXrefSpec>)>();
+            var externalXrefSpec = new List<(string, Lazy<ExternalXrefSpec>)>();
+            var externalXref = new List<ExternalXref>();
+
             var content = File.ReadAllBytes(filePath);
 
             // TODO: cache this position mapping if xref map file not updated, reuse it
-            var (xrefSpecPositions, repositoryUrl) = GetXrefSpecPosAndRepoUrl(content, filePath);
+            var (xrefSpecPositions, xrefPositions, repositoryUrl, docsetName) = GetXrefSpecPosXrefPosAndRepoUrl(content, filePath);
 
             foreach (var (uid, start, end) in xrefSpecPositions)
             {
-                result.Add((uid, new Lazy<ExternalXrefSpec>(() =>
+                externalXrefSpec.Add((uid, new Lazy<ExternalXrefSpec>(() =>
                 {
                     using var stream = File.Open(filePath, FileMode.Open, FileAccess.Read, FileShare.Read);
                     var spec = JsonUtility.DeserializeData<ExternalXrefSpec>(ReadJsonFragment(stream, start, end), new FilePath(filePath));
                     spec.RepositoryUrl = repositoryUrl;
+                    spec.DocsetName = docsetName;
                     return spec;
                 })));
             }
-            return result;
+
+            foreach (var (start, end) in xrefPositions)
```

Comment:
add a todo we are going to remove the optimization in another PR

Reply:
OK.